src/NuGet.Clients/NuGet.Credentials/CredentialService.cs

```diff
@@ -0,0 +1,191 @@
+        /// New CredentialService objects will be populated with these default
+        /// configured providers.
+        /// </summary>
+        public static Lazy<IEnumerable<ICredentialProvider>> DefaultProviders
+        {
+            set
+            {
+                _defaultProviders = value;
+            }
+            internal get { return _defaultProviders;}
+        }
+
+        /// <summary>
+        /// Gets the currently configured providers.
+        /// Internal set is provided for use by unit tests.
+        /// </summary>
+        public IEnumerable<ICredentialProvider> Providers
+        {
+            get { return _providerList ?? _defaultProviders.Value; }
+            internal set { _providerList =  value; }
+        } 
+
+        /// <summary>
+        /// Provides credentials for http requests.
+        /// </summary>
+        /// <param name="uri">The uri of a web resource for which credentials are needed.</param>
+        /// <param name="proxy">The currently configured proxy. It may be necessary for CredentialProviders
+        /// to use this proxy in order to acquire credentials from their authentication source.</param>
+        /// <param name="isProxy">If true, get credentials to authenticate for the requested proxy.
+        /// If false, the credentials are intended for a remote service.</param>
+        /// <param name="cancellationToken">A cancellation token.</param>
+        /// <returns>A credential object, or null if no credentials could be acquired.</returns>
+        public async Task<ICredentials> GetCredentials(Uri uri, IWebProxy proxy, bool isProxy,
+            CancellationToken cancellationToken)
+        {
+            if (uri == null)
+            {
+                throw new ArgumentNullException(nameof(uri));
+            }
+
+            ICredentials response = null;
+
+            foreach (var provider in Providers)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var retryKey = RetryCacheKey(uri, isProxy, provider);
+                var isRetry = _retryCache.ContainsKey(retryKey);
+
+                try
+                {
+                    // This local semaphore ensures one provider active per process.
+                    // We can consider other ways to allow more concurrency between providers, but need to
+                    // ensure that only one interactive dialogue is ever presented at a time, and that
+                    // providers are not writing shared resources.
+                    // Since this service is called only when cached credentials are not available to the caller,
+                    // such an optimization is likely not necessary.
+                    _providerSemaphore.WaitOne();
+
+                    if (!TryFromCredentialCache(uri, isProxy, isRetry, provider, out response))
+                    {
+                        response = await provider.Get(uri, proxy, isProxyRequest: isProxy, isRetry: isRetry,
```

Comment:
Change the response to CredentialsProviderResult instead of throwing exceptions.
the result object should have the following properties Status, ICredentials, Message